libdnf/base/transaction.cpp

```diff
@@ -323,13 +273,28 @@ TransactionPackage Transaction::Impl::make_transaction_package(
     Id id,
     TransactionPackage::Action action,
     rpm::solv::GoalPrivate & solved_goal,
-    std::map<Id, std::vector<Id>> & replaced) {
+    std::map<Id, std::vector<Id>> & replaced,
+    rpm::PackageQuery installed_query) {
     auto obs = solved_goal.list_obsoleted_by_package(id);
 
-    libdnf_assert(!obs.empty(), "No obsoletes for {}", transaction_item_action_to_string(action));
-
     rpm::Package new_package(base, rpm::PackageId(id));
-    auto reason = new_package.get_reason();
+
+    transaction::TransactionItemReason reason;
+    if (action == TransactionPackage::Action::INSTALL) {
+        reason = std::max(new_package.get_reason(), solved_goal.get_reason(id));
+
+        installed_query.filter_name({new_package.get_name()});
+        installed_query.filter_arch({new_package.get_arch()});
+        // For installonly packages: if the NA is already on the system, but
+        // not recorded in system state as installed, it was installed outside
+        // DNF and we want to preserve NONE as reason
+        if (!installed_query.empty() && new_package.get_reason() == transaction::TransactionItemReason::NONE) {
```

Comment:
If we will move the logic to `new_package.get_reason()` we can remove above lines. What do you think?

Reply:
Anyway this line is somehow redundant because it reverts decision from line 284 using the same set of information plus result from the mandatory `installed_query`.